mesa_geo/raster_layers.py

```diff
@@ -0,0 +1,476 @@
+        super().__init__(crs)
+        self._width = width
+        self._height = height
+        self._total_bounds = total_bounds
+        self._update_transform()
+
+    @property
+    def width(self) -> int:
+        return self._width
+
+    @width.setter
+    def width(self, width: int) -> None:
+        self._width = width
+        self._update_transform()
+
+    @property
+    def height(self) -> int:
+        return self._height
+
+    @height.setter
+    def height(self, height: int) -> None:
+        self._height = height
+        self._update_transform()
+
+    @property
+    def total_bounds(self) -> np.ndarray:
+        return self._total_bounds
+
+    @total_bounds.setter
+    def total_bounds(self, total_bounds: np.ndarray) -> None:
+        self._total_bounds = total_bounds
+        self._update_transform()
+
+    @property
+    def transform(self) -> Affine:
+        return self._transform
+
+    @property
+    def resolution(self) -> Tuple[float, float]:
+        """
+        Returns the (width, height) of a cell in the units of CRS.
+        """
+        a, b, _, d, e, _, _, _, _ = self.transform
+        return math.sqrt(a**2 + d**2), math.sqrt(b**2 + e**2)
+
+    def _update_transform(self) -> None:
+        self._transform = rio.transform.from_bounds(
+            *self.total_bounds, width=self.width, height=self.height
+        )
+
+    def to_crs(self, crs, inplace=False) -> RasterBase | None:
+        raise NotImplementedError
+
+    def out_of_bounds(self, pos: Coordinate) -> bool:
+        x, y = pos
+        return x < 0 or x >= self.width or y < 0 or y >= self.height
+
+
+class Cell(Agent):
+    pos: Coordinate | None  # (x, y), origin is at lower left corner of the grid
+    indices: Coordinate | None  # (row, col), origin is at upper left corner of the grid
+
+    def __init__(self, pos=None, indices=None):
+        super().__init__(None, None)
+        self.pos = pos
+        self.indices = indices
+
+    def step(self):
+        pass
+
+
+class RasterLayer(RasterBase):
+    cells: List[List[Cell]]
+    _neighborhood_cache: Dict[Any, List[Coordinate]]
+
+    def __init__(self, width, height, crs, total_bounds, cell_cls=Cell):
+        super().__init__(width, height, crs, total_bounds)
+        self.cell_cls = cell_cls
+        self.cells = []
+        for x in range(self.width):
+            col: List[cell_cls] = []
+            for y in range(self.height):
+                row_idx, col_idx = self.height - y - 1, x
+                col.append(self.cell_cls(pos=(x, y), indices=(row_idx, col_idx)))
+            self.cells.append(col)
+
+        self._neighborhood_cache = {}
+
+    @overload
+    def __getitem__(self, index: int) -> List[Cell]:
+        ...
+
+    @overload
+    def __getitem__(self, index: Tuple[int | slice, int | slice]) -> Cell | List[Cell]:
+        ...
+
+    @overload
+    def __getitem__(self, index: Sequence[Coordinate]) -> List[Cell]:
+        ...
+
+    def __getitem__(
```

Comment:
Is the code in this method copied almost verbatim from `mesa.space.Grid`? If so, we should inform in the docstring about it. But there are potential new changes to it in https://github.com/projectmesa/mesa/pull/815. Having to manually sync between them would not be ideal.

Reply:
You're right! Initially I planned to make RasterLayer derived from Grid:

```python
class RasterLayer(GeoBase, mesa.space.Grid):
    ...
```

But then I don't want it to be able to hold agents, since we want our users to add agents directly into GeoSpace.

I'm also aware of the new changes, but they are not yet released. I hope such changes won't happen too often in the future so that manual sync remains feasible.

Reply:
We could define a common class in Mesa that has this method. Is `__getitem__` the only one that is copied over so far?

Reply:
I'm afraid not. What's copied over so far:

```
__getitem__
__iter__
coord_iter
iter_neighborhood
get_neighborhood
iter_neighbors
get_neighbors  # copied and renamed to `get_neighboring_cells`
out_of_bounds  # not yet copied but could be useful
iter_cell_list_contents
get_cell_list_contents
```

Methods that are not copied over:

```
torus_adj
neighbor_iter
move_agent
place_agent
_place_agent
remove_agent
is_cell_empty
move_to_empty
find_empty
exists_empty_cells
```

Another difference is that `Grid` has `self.grid: List[List[Agent | None]]`, whereas it is `self.cells: List[List[Cell]]` in `RasterLayer`.

Reply:
That's a lot of methods. We should definitely replace the copying into inheritance very soon. I'm fine with it for now though.

Reply:
That's indeed a lot of methods. Inheritance would make it much better, but I would not want to mess up the classes in Mesa either.